pbp/publickey.py

```diff
@@ -246,24 +254,44 @@ def buffered_verify(infd, outfd, basedir, self = None):
     if sender and hashsum == hashsum1:
         return sender
 
+def getkey(size, pwd='', empty=False, text=''):
+    # queries the user twice for a passphrase if neccessary, and
+    # returns a scrypted key of length size
+    # allows empty passphrases if empty == True
+    # 'text' will be prepended to the password query
+    # will not query for a password if pwd != ''
+    #clearpwd = (pwd.strip()=='')
+    pwd2 = not pwd
+    if not pwd:
+        while pwd != pwd2 or (not empty and not pwd.strip()):
+            pwd = getpass.getpass('1/2 %s Passphrase: ' % text)
+            if pwd.strip():
+                pwd2 = getpass.getpass('2/2 %s Repeat passphrase: ' % text)
+    #if isinstance(pwd2, str):
+    #   clearmem(pwd2)
+    if pwd.strip():
+        key = scrypt.hash(pwd, scrypt_salt)[:size]
+        #if clearpwd: clearmem(pwd)
+        return key
+
 def get_public_keys(basedir=None):
-    if not basedir: basedir=pbp.defaultbase
+    if not basedir: basedir=defaultbase
     basedir=os.path.expandvars(os.path.expanduser(basedir))
     pk_dir = get_pk_dir(basedir)
     if not os.path.exists(pk_dir):
         return
-    for root, ext in imap(os.path.splitext, os.listdir(pk_dir)):
+    for root, ext in map(os.path.splitext, os.listdir(pk_dir)):
```

Comment:
i prefer an iterator here

Reply:
This is an iterator still in both Python 2 and Python 3, note the import at the top of the file. `imap` is still imported but renamed to `map`, and `map` is an iterator by default in Python 3.